Ground work for identify to host

ralexstokes opened the pull request:
Builds on top of #334.

ralexstokes commented:
i'll rebase after merging in #334.

mhchia commented on tests/factories.py:
```diff
@@ -74,19 +76,27 @@ class Meta:
+        )
+        return tuple(
+            BasicHost(key_pair.public_key, swarm)
+            for swarm, key_pair in range(swarms_and_keys)
```
Really nice catch! Nit: there is an extra `:`.

ralexstokes replied:
yeah -- this was the WIP part of this PR! the interpreter would have not liked this had it tried to interpret it

ralexstokes commented on libp2p/host/basic_host.py:
```diff
@@ -44,14 +45,16 @@ class BasicHost(IHost):
         network: INetwork,
         default_protocols: "OrderedDict[TProtocol, StreamHandlerFn]" = None,
     ) -> None:
+        self._public_key = public_key
```
nice catch

ralexstokes commented on libp2p/host/defaults.py:
```diff
@@ -1,11 +1,11 @@
 
-def get_default_protocols() -> "OrderedDict[TProtocol, StreamHandlerFn]":
-    return DEFAULT_HOST_PROTOCOLS.copy()
+def get_default_protocols(host: IHost) -> "OrderedDict[TProtocol, StreamHandlerFn]":
```
exactly -- in a future PR, we will have the `identify` protocol that requests an `IHost`. local development ran a bit ahead of this PR ;)

ralexstokes commented on libp2p/protocol_muxer/multiselect_muxer_interface.py:
```diff
@@ -23,6 +23,9 @@ def add_handler(self, protocol: TProtocol, handler: StreamHandlerFn) -> None:
         :param handler: handler function
         """
 
+    def get_protocols(self) -> Tuple[TProtocol, ...]:
```
in this case no because we can write the implementation for all such `IMultiselectMuxer` using only things declared in this interface.

it would be different if we didn't commit to `handlers: Dict[TProtocol, StreamHandlerFn]` here

ralexstokes commented:
@NIC619 @mhchia i've addressed your feedback and give you another chance for final review -- if there is nothing then we can go ahead and merge this and in the meantime i'll start on the PR to tie all of this stuff together
